notebooks/02_model/mmlspark_lightgbm_criteo.ipynb

```diff
@@ -0,0 +1,407 @@
+      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'azure.mgmt.authorization'"
+     ]
+    }
+   ],
+   "source": [
+    "from azureml.core import Workspace\n",
+    "from azureml.core.model import Model\n",
+    "\n",
+    "print(\"Azure ML SDK version: \", azureml.core.VERSION)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data Preparation\n",
+    "The Criteo Display Advertising Challenge (DAC) dataset [4] is a well-known industry benchmarking dataset for developing CTR prediction models, and is used frequently by research papers. The original dataset contains over 45M rows, but there is also a down-sampled dataset which has 100,000 rows (this can be used by setting DATA_SIZE = 'sample').<br><br>\n",
+    "The dataset contains 1 label column and 38 feature columns, where 13 columns are integer values (int00-int12) and 25 columns are categorical features (cat00-cat24).<br><br>\n",
+    "What the columns represent is not provided, but for this case we can consider the integer and categorical values as features representing the user and / or item content. The label is binary and indicates a user interaction with an item, so this is a useful dataset to demonstrate how to build a model that will predict likelihood of a user interacting with an item based on the user and item content features.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {
+    "tags": [
+     "parameters"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "DATA_SIZE = 'sample'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "8.79MB [00:02, 3.62MB/s]                            \n"
+     ]
+    }
+   ],
+   "source": [
+    "raw_data = load_spark_df(size=DATA_SIZE, spark=spark)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Feature Processing\n",
+    "The feature data provided has many missing values across both integer and categorical feature fields. In addition the categorical features have many distinct values, so effectively cleaning and representing the feature data is an important step prior to training a model.<br><br>\n",
+    "One of the simplest ways of managing both features that have missing values as well as high cardinality is to use the hashing trick. The FeatureHasher transformer will pass integer values through and will hash categorical features into a sparse vector of lower dimensionality which can be used effectively by LightGBM.<br><br>\n",
+    "First the dataset is split randomly for training and testing and feature processing is applied to each dataset."
```

Comment:
I am wondering if splitting should happen before hashing or after...

Reply:
Good point. Splitting should happen after hashing. Now we will end up with inconsistencies between train and test sets.

Reply:
i disagree, it's good practice to fit your feature transformer on the training data only to avoid leakage, so processing must happen after splitting. In this case there is no fitting of the hash function, and it is applied uniformly for train and test.

Reply:
You are right. This transformer does not depend on the data.

Reply:
@gramhagen I see your point. This makes sense.